src/net.cpp

```diff
@@ -420,6 +420,26 @@ CNode* ConnectNode(CAddress addrConnect, const char *pszDest, bool fConnectToMas
             return NULL;
         }
 
+        if (pszDest && addrConnect.IsValid()) {
+            // It is possible that we already have a connection to the IP/port pszDest resolved to.
+            // In that case, drop the connection that was just created, and return the existing CNode instead.
+            // Also store the name we used to connect in that CNode, so that future FindNode() calls to that
+            // name catch this early.
+            CNode* pnode = FindNode((CService)addrConnect);
+            if (pnode)
+            {
+                pnode->AddRef();
```

Comment:
`AddRef` for ("normal") existing connection in bitcoin was a bug https://github.com/dashpay/dash/pull/924 https://github.com/bitcoin/bitcoin/pull/8372. Using `AddRef` in situations where we select node with existing connection as a masternode to connect to, allowed us to track some related bugs iirc. These `AddRef`s and corresponding `Release` in `ThreadSocketHandler` for `fMasternode` flag are probably not needed anymore, so I _guess_ they can be safely removed (in both places simultaniously ofc).

Also, I think triggering `fMasternode` flag can be simplified even further:
```
pnode->fMasternode |= fConnectToMasternode;
```